### ISSUE TITLE
Moving from 4.1.3 to 4.2.0 -> xmlhttprequest-ssl' relative to engine.io-client error

### ISSUE THREAD

**Issue by hardcodet** [socketio/socket.io-client#1494]
**Describe the bug**

I was toying around with socket.io in codesandbox.io, and experienced a crash pretty much right away. When switching to 4.1.3, the error went away:

_ModuleNotFoundError
Could not find module in path: 'xmlhttprequest-ssl' relative to '/node_modules/engine.io-client/lib/transports/index.js'_

**To Reproduce**

* codesandbox.io/
* add a reference to 4.2.0 and initialize a socket -> will crash
* downgrade to 4.1.3 -> workd

Please fill the following code example:

Socket.IO server version: `x.y.z`

*Client (4.2.0)*

```js
mport { io, Socket } from "socket.io-client";
import { config } from "./config";

export class SocketClient {
  socket: Socket;

  constructor() {
    this.socket = io("http://localhost:3000", {});
  }
}

```

**Comment by darrachequesne** [comment]
We used a fork of `xmlhttprequest-ssl`, until the changes were merged upstream. We have removed our fork in `socket.io-client@4.2.0`, which may explain (or at least, be linked to) your issue: https://github.com/socketio/engine.io-client/commit/2de8b4ef97623c65f772710aa142a63c3bbbf1d7

But switching version should properly resolve the dependencies, so I guess it's a problem with codesandbox?

**Comment by alpha-titan** [comment]
There are some bugs in the `4.2.0`  client, I am not able to connect to the server.

I have **enabled** CORS on the **socket server** as well as the **express server**

I was stuck at this issue for hours tried everything no luck establish a WebSocket connection with the server, finally downgraded to `4.1.3`

The below code works on `4.1.3` and doesn't work on `4.2.0`

```
const [socket, setSocket] = useState();
const handleSocket = () => {
        newSocket = io("ws://localhost:4000");
	newSocket.connect();
	console.log(newSocket.connected);
        setSocket(newSocket);
}
useEffect(()=>{
         handleSocket();
},[setSocket])
```
Upon inspecting Network for ws connections

using `4.1.3` connection is created

<img width="734" alt="Screenshot 2021-09-20 at 11 11 07 PM" src="https://user-images.githubusercontent.com/31860431/134048989-499b5309-2b5c-41cf-b58e-9192ee6cc22e.png">

using `4.2.0` pending 

<img width="725" alt="Screenshot 2021-09-20 at 11 13 34 PM" src="https://user-images.githubusercontent.com/31860431/134049132-17b87186-c28e-447b-9177-6517d51e5894.png">


when I use the `CDN` and comment out the import it works again.

**Comment by hardcodet** [comment]
@darrachequesne  Could well be a codesandbox issue - I just did a quick hello world there, but wanted to report it in case it would be of interest to you.

OTOH, I would assume that codesandbox generically just downloads packages as specified in package.json, so I'm not sure it _could_ be buggy there and make any assumptions (in the end, it did end up with code that was referring to that SSL library).

@alpha-titan Looks like your issue is not related to the one reported here?

**Comment by darrachequesne** [comment]
@alpha-titan I was not able to reproduce, could you please check https://github.com/socketio/socket.io-fiddle/tree/issues/socket.io/1494? Could you please open a new issue with all necessary details?

@hardcodet I could in fact reproduce:

![screenshot](https://user-images.githubusercontent.com/13031701/134119870-5c49677f-04e9-4224-bf71-aaa75ad4f96d.png)

I don't understand why though. A fresh "npm install" or "yarn add" both includes the `xmlhttprequest-ssl` package:

```
$ npm ls xmlhttprequest-ssl
/test
└─┬ socket.io-client@4.2.0
  └─┬ engine.io-client@5.2.0
    └── xmlhttprequest-ssl@2.0.0
```

If someone can shed some light on this...

**Comment by alpha-titan** [comment]
@darrachequesne  I have a React` client (created with vite.js) with version socket `4.2.0` and an `express server` running on different `port` and I was not possible to establish a connection.

**Comment by alpha-titan** [comment]
@hardcodet Yeah sorry for that I will open a new Issue

**Comment by louisgv** [comment]
I encountered this while upgrading to 4.2.0 in react-native environment. The metro bundler gave this error:

> Unable to resolve module ../../contrib/xmlhttprequest-ssl/XMLHttpRequest from P:\Projects\considr\ponde
> r\native\node_modules\engine.io-client\lib\transports\index.js:
> 
> None of these files exist:
>   * node_modules\engine.io-client\contrib\xmlhttprequest-ssl\XMLHttpRequest(.native|.ios.ts|.native.ts|
> .ts|.ios.tsx|.native.tsx|.tsx|.ios.js|.native.js|.js|.ios.jsx|.native.jsx|.jsx|.ios.json|.native.json|.
> json)
>   * node_modules\engine.io-client\contrib\xmlhttprequest-ssl\XMLHttpRequest\index(.native|.ios.ts|.nati
> ve.ts|.ts|.ios.tsx|.native.tsx|.tsx|.ios.js|.native.js|.js|.ios.jsx|.native.jsx|.jsx|.ios.json|.native.
> json|.json)
> > 1 | const XMLHttpRequest = require("xmlhttprequest-ssl");
>   2 | const XHR = require("./polling-xhr");
>   3 | const JSONP = require("./polling-jsonp");
>   4 | const websocket = require("./websocket");

Doing npm ls show the package was installed:
```
$ npm ls xmlhttprequest-ssl
`-- socket.io-client@4.2.0
  `-- engine.io-client@5.2.0
    `-- xmlhttprequest-ssl@2.0.0
```

**Comment by louisgv** [comment]
It seems the core issue is that somehow the bundler is trying to resolve the package relatively? Based on the log above, it's tring to find the package in `../../contrib/xmlhttprequest-ssl`

**Comment by louisgv** [comment]
Very weird - I grep through the installed code in `node_modules` for `contrib` and nothing was found. I have reinstall my dependencies a couple time and still seeing the issue.

**Comment by darrachequesne** [comment]
@louisgv did you try to reset metro-s cache with `--reset-cache`?

Reference: https://facebook.github.io/metro/docs/troubleshooting

**Comment by jsjoeio** [comment]
I'm seeing a similar issue. I landed on [this CodeSandbox](https://codesandbox.io/s/web-terminal-turorial-client-sgjgg?file=/src/index.js) from [this article](https://dev.to/saisandeepvaddi/how-to-create-web-based-terminals-38d).

![image](https://user-images.githubusercontent.com/3806031/135765793-74ffee41-6bb8-459e-a4d3-b17a229bb13d.png)

Updating `"socket.io-client": "2.3.0"` -> `4.2.0` results in the same error. I wonder if this is specific to CodeSandbox 🤔

**Comment by jsjoeio** [comment]
@darrachequesne I found this: https://github.com/codesandbox/codesandbox-client/issues/4456#issuecomment-846326066

**Comment by darrachequesne** [comment]
This should be fixed by https://github.com/socketio/engine.io-client/commit/49719142f65e23efa65fca4f66765ded5d955972, included in version `4.3.1`.

Please reopen if needed.
